### ISSUE TITLE
mpl: memory tracing

### ISSUE THREAD

**Issue by raffenet** [pmodels/mpich#4329]
I propose to remove internal memory tracing functionality in MPL (trmem). https://github.com/google/sanitizers/wiki/AddressSanitizer supports leak and OOB access detection (and more!) well and without dramatic performance penalty so that it can be run regularly in Jenkins (https://pmrs-jenkins.cels.anl.gov/job/mpich-master-address-sanitizer/). Note some timeout tweaks are still needed, but the overall results are much more workable than valgrind testing.

I am not proposing to remove MPL memory allocation statistics, or object handle allocation tracing. Thoughts?

**Comment by wesbland** [comment]
I don't think I understand what you're proposing. Would one have to do something different to get memory tracing at this point? It's still helpful to have the warning when you have un-freed memory. If doing so requires an external tool, it's less likely to happen.

**Comment by raffenet** [comment]
Yes, you would have to build with the address sanitizer compiler options. We could add a configuration for it. There's already a Jenkins job.

However after I created this, I realized removing this code wouldn't be useful. Since we added statistics, converting `MPL_malloc` -> `malloc` is a no-go. The only benefit it would give is if another allocation function were to be added, a tracing version would not need to be implemented along side it. See `MPL_traligned_alloc` for the last example. Since I don't foresee any new allocation functions coming soon, I think we can just drop this idea.
